src/charm.py

```diff
@@ -170,6 +199,44 @@ def _run_sql_migration(self, set_timeout: bool) -> None:
         stdout, _ = process.wait_output()
         logger.info(f"Executing automigration: {stdout}")
 
+    def _oauth_relation_peer_data_key(self, relation_id: int) -> str:
+        return f"oauth_{relation_id}"
+
+    @property
+    def _peers(self):
+        """Fetch the peer relation."""
+        return self.model.get_relation(PEER)
+
+    def _set_peer_data(self, key: str, data: Dict) -> None:
+        """Put information into the peer data bucket instead of `StoredState`."""
+        self._peers.data[self.app][key] = json.dumps(data)
+
+    def _get_peer_data(self, key: str) -> Dict:
+        """Retrieve information from the peer data bucket instead of `StoredState`."""
+        if not self._peers:
+            return {}
+        data = self._peers.data[self.app].get(key, "")
+        return json.loads(data) if data else {}
+
+    def _pop_peer_data(self, key: str) -> Dict:
+        """Retrieve information from the peer data bucket."""
+        if not self._peers:
+            return {}
+        data = self._peers.data[self.app].pop(key, "")
+        return json.loads(data) if data else {}
+
+    def _set_oauth_relation_peer_data(self, relation_id: int, data: Dict) -> None:
+        key = self._oauth_relation_peer_data_key(relation_id)
+        self._set_peer_data(key, data)
+
+    def _get_oauth_relation_peer_data(self, relation_id: int) -> None:
+        key = self._oauth_relation_peer_data_key(relation_id)
+        return self._get_peer_data(key)
+
+    def _pop_oauth_relation_peer_data(self, relation_id: int) -> None:
```

Comment:
I agree that if this fails, it could cause unexpected errors to happen.

BUT going through all the relations and verifying that the corresponding client exists can be a very expensive operation, depending on how the admin chooses to use the platform. We would have to get the entire client table for the database and compare that with the contents of the juju relation.

I think that the best we can do is:
- Provide a sanitization command that would do that checking.
- Catch any errors that may be thrown when popping from the peer relation and log a descriptive error, urging the admin to run the sanitization command.

Reply:
After some more thought on this, I realized that, while this is a very valid concern, in general we don't handle failures when trying to get/set relation data (correct me if I'm wrong). I don't see why this case is any different.

To fix this we could wrap this in a `try/except` block and fire an event for reconciling the peer data, but we need to first understand when that can actually fail and what it means that it failed (will juju try to fire the event again?) and if we actually need the user to do something or not (can be just a temporary error?).

Reply:
Taking a step back on this one I consider deleting clients when a relation is broken as something that is _nice to have_ form a security/operational point of view, but not _essential_ (unless the prior credentials are leaked)

For the time being, and until we are able to observe the real behaviour of the charm in the field, I suggest that in the event of an error we throw an informative message that instructs the admin to check the active clients in hydra, either through the charm actions or the cli

I would also open an issue on this repo detailing the possible error conditions and further work required